README.md
<p align="center">
  <img src="https://www.corda.net/wp-content/uploads/2016/11/fg005_corda_b.png" alt="Corda" width="500">
</p>

# financial_derivatives_demo

Welcome to the Financial Derivatives Demo in Corda.  

This is for demonstration purposes only, and it comes as-is, open source with no warantees of any kind.  Do not use this pricing information for the purposes of real-world valuations. I cant stress enough that this is all pretty approximate, and not to be used as is in a real world financial application.
If you ignore those warning, and come to regret it, please don't sue me. 

This project is built on top of the Cordapp Template in Kotlin, so the stucture is borrowed from there. 

Critical components here beyond Corda are as follows.

1) QuantLib - Quantlib is a free open source library for quantitative finance written in C++ by people way smarter than me. 
I'm greatful to the community around it, find more info here: https://www.quantlib.org/

2) SWIG (Simplified Wrapper and Interface Generator) - Because the quantlib library is written in C++ in order to access it from a Java context we require swig bindings.  This is also a wonderful opensource library, see details here:   http://www.swig.org/

3) Quandl - The hope eventually is to build against multiple market data sources, Boomberg, Reuters etc. for now building using the freely available API from Quandl and some treasury data. Quandl info here https://www.quandl.com/

Quandl is conveniently repackaging US treasury data found here: 
https://www.treasury.gov/resource-center/data-chart-center/interest-rates/pages/textview.aspx?data=yield

The rest of this is doc included with the original COrDapp template... I will clean it up in time  

# CorDapp Template - Kotlin

Welcome to the Kotlin CorDapp template. The CorDapp template is a stubbed-out CorDapp that you can use to bootstrap 
your own CorDapps.

**This is the Kotlin version of the CorDapp template. The Java equivalent is 
[here](https://github.com/corda/cordapp-template-java/).**

# Pre-Requisites

See https://docs.corda.net/getting-set-up.html.

# Usage

## Running tests inside IntelliJ

We recommend editing your IntelliJ preferences so that you use the Gradle runner - this means that the quasar utils
plugin will make sure that some flags (like ``-javaagent`` - see below) are
set for you.

To switch to using the Gradle runner:

* Navigate to ``Build, Execution, Deployment -> Build Tools -> Gradle -> Runner`` (or search for `runner`)
  * Windows: this is in "Settings"
  * MacOS: this is in "Preferences"
* Set "Delegate IDE build/run actions to gradle" to true
* Set "Run test using:" to "Gradle Test Runner"

If you would prefer to use the built in IntelliJ JUnit test runner, you can run ``gradlew installQuasar`` which will
copy your quasar JAR file to the lib directory. You will then need to specify ``-javaagent:lib/quasar.jar``
and set the run directory to the project root directory for each test.

## Running the nodes

See https://docs.corda.net/tutorial-cordapp.html#running-the-example-cordapp.

## Interacting with the nodes

### Shell

When started via the command line, each node will display an interactive shell:

    Welcome to the Corda interactive shell.
    Useful commands include 'help' to see what is available, and 'bye' to shut down the node.
    
    Tue Nov 06 11:58:13 GMT 2018>>>

You can use this shell to interact with your node. For example, enter `run networkMapSnapshot` to see a list of 
the other nodes on the network:

    Tue Nov 06 11:58:13 GMT 2018>>> run networkMapSnapshot
    [
      {
      "addresses" : [ "localhost:10002" ],
      "legalIdentitiesAndCerts" : [ "O=Notary, L=London, C=GB" ],
      "platformVersion" : 3,
      "serial" : 1541505484825
    },
      {
      "addresses" : [ "localhost:10005" ],
      "legalIdentitiesAndCerts" : [ "O=PartyA, L=London, C=GB" ],
      "platformVersion" : 3,
      "serial" : 1541505382560
    },
      {
      "addresses" : [ "localhost:10008" ],
      "legalIdentitiesAndCerts" : [ "O=PartyB, L=New York, C=US" ],
      "platformVersion" : 3,
      "serial" : 1541505384742
    }
    ]
    
    Tue Nov 06 12:30:11 GMT 2018>>> 

You can find out more about the node shell [here](https://docs.corda.net/shell.html).

### Client

`clients/src/main/kotlin/com/template/Client.kt` defines a simple command-line client that connects to a node via RPC 
and prints a list of the other nodes on the network.

#### Running the client

##### Via the command line

Run the `runTemplateClient` Gradle task. By default, it connects to the node with RPC address `localhost:10006` with 
the username `user1` and the password `test`.

##### Via IntelliJ

Run the `Run Template Client` run configuration. By default, it connects to the node with RPC address `localhost:10006` 
with the username `user1` and the password `test`.

### Webserver

`clients/src/main/kotlin/com/template/webserver/` defines a simple Spring webserver that connects to a node via RPC and 
allows you to interact with the node over HTTP.

The API endpoints are defined here:

     clients/src/main/kotlin/com/template/webserver/Controller.kt

And a static webpage is defined here:

     clients/src/main/resources/static/

#### Running the webserver

##### Via the command line

Run the `runTemplateServer` Gradle task. By default, it connects to the node with RPC address `localhost:10006` with 
the username `user1` and the password `test`, and serves the webserver on port `localhost:10050`.

##### Via IntelliJ

Run the `Run Template Server` run configuration. By default, it connects to the node with RPC address `localhost:10006` 
with the username `user1` and the password `test`, and serves the webserver on port `localhost:10050`.

#### Interacting with the webserver

The static webpage is served on:

    http://localhost:10050

While the sole template endpoint is served on:

    http://localhost:10050/templateendpoint
    
# Extending the template

You should extend this template as follows:

* Add your own state and contract definitions under `contracts/src/main/kotlin/`
* Add your own flow definitions under `workflows/src/main/kotlin/`
* Extend or replace the client and webserver under `clients/src/main/kotlin/`

For a guided example of how to extend this template, see the Hello, World! tutorial 
[here](https://docs.corda.net/hello-world-introduction.html).
# financial_derivatives_demo


quantlib_swig_bindings/ChangeLog.txt
commit 96d1506e63755b5700095083af4a862a73754903
Merge: fdcb4df 3d21ca3
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Thu, 18 Jul 2019 12:52:43 +0200

    Merge pull request #193.

commit 3d21ca357d05600f273ff988dac1ba34794874cb
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Wed, 17 Jul 2019 18:15:57 +0200

    Avoid deprecation warning.

 SWIG/volatilities.i | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

commit fdcb4dff4b4d2b419e371964269e7f7e71b211b9
Merge: a66f494 3e366b0
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Wed, 17 Jul 2019 11:09:26 +0200

    Merge pull request #192.

commit 3e366b0df02c4ca837cecb344f1f558f24a76ffb
Author: klausspanderen <klaus@spanderen.de>
Date:   Tue, 16 Jul 2019 21:25:18 +0200

    added test for HestonSLV DoubleBarrier option pricing

 Python/test/slv.py | 58 +++++++++++++++++++++++++++++++++++++++++++-----------
 1 file changed, 47 insertions(+), 11 deletions(-)

commit a66f4944365f2b2c95f3f48becf853aa1e6bc4e0
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Tue, 16 Jul 2019 12:00:29 +0200

    Update VC# project.

 CSharp/csharp/NQuantLib.csproj | 27 ++++++++++++++++++++++++++-
 1 file changed, 26 insertions(+), 1 deletion(-)

commit d5fe830f602c68725e730fcbce96dcf0cf31b06f
Merge: 330c6cb 6c70cee
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Fri, 12 Jul 2019 18:17:31 +0200

    Merge pull request #185.

commit 330c6cbe571cd3716383c07c342db961afd9826d
Merge: a6a1592 1e27d24
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Fri, 12 Jul 2019 11:12:36 +0200

    Merge pull request #190.

commit 6c70cee882a09c13b4e3f4c093ccf5cc3e075c16
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Fri, 12 Jul 2019 10:16:04 +0200

    Revert unmodified file.

 SWIG/slv.i | 1 -
 1 file changed, 1 deletion(-)

commit c585fbb10075b023d55f117c16cc103631a6cca1
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Fri, 12 Jul 2019 10:13:03 +0200

    Remove duplicated class.

 SWIG/slv.i | 48 ------------------------------------------------
 1 file changed, 48 deletions(-)

commit a6a15928197ce53a95e0e91c29f4fd138e32f894
Merge: 02605ec 906847b
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Thu, 11 Jul 2019 17:48:01 +0200

    Merge pull request #182.

commit 02605ec7504f7f2eebc270696b6dd87e38a7e289
Merge: 90e62d6 65124e0
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Thu, 11 Jul 2019 17:33:44 +0200

    Merge pull request #181.

commit 1e27d24e6f394084befcee44c5dc513c4b97875a
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Thu, 11 Jul 2019 15:18:21 +0200

    Export G2 processes.

 SWIG/stochasticprocess.i | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

commit 90e62d60ad916f80a652f5087de8192a9b4b7dbc
Merge: f397282 c8b62f2
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Thu, 11 Jul 2019 13:20:10 +0200

    Merge pull request #158.

commit f39728205a431baa4d9867158ddd3ce70c8922f5
Merge: 0d21bd2 a6b4e53
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Thu, 11 Jul 2019 12:13:03 +0200

    Merge pull request #175.

commit 0d21bd2d01075134015c32fd84c05cb3a42d736e
Merge: ce80e1f f8edd16
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Thu, 11 Jul 2019 10:57:40 +0200

    Merge pull request #169.

commit ce80e1f47e1fef78a7f9dfc9f791b111abc144e4
Merge: 256c19e 087149c
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Wed, 10 Jul 2019 13:10:50 +0200

    Merge pull request #174.

commit f8edd160556a7e49e17c1bcd48fb429bb39e8c11
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Wed, 10 Jul 2019 11:32:18 +0200

    Remove C++11-only syntax.

 .gitignore      |  7 +------
 SWIG/capfloor.i | 12 ++++--------
 2 files changed, 5 insertions(+), 14 deletions(-)

commit 256c19e7e176fc0e4e2f8af378682e5a88c1f79a
Merge: 993f4e4 b6d8639
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Tue, 9 Jul 2019 17:17:21 +0200

    Merge pull request #155.

commit 5945125618b667aa8651ee54e81d860668f1575a
Author: Matthias Lungwitz <lungwitz@gmail.com>
Date:   Fri, 14 Jun 2019 00:05:56 +0200

    HestonBlackVolSurface

 SWIG/slv.i | 49 +++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 49 insertions(+)

commit c8b62f2fd7f0290eedbcf368451674f1bbda48ee
Author: Luigi Ballabio <luigi.ballabio@gmail.com>
Date:   Mon, 10 Jun 2019 15:06:01 +0200

    Work around invalid generated code in C# module.

 SWIG/volatilities.i | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

commit 906847b81e3bf99d4a27f5e021876b6554656759
Author: Peter Caspers <peter.caspers@quaternion.com>
Date:   Fri, 7 Jun 2019 18:53:26 +0200

    improve description (thanks to Giuseppe Trapani)

 Python/examples/gaussian1d-models.py | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

commit 65124e0746f83dc0109f2463b413974d3193289a
Author: Matthias Lungwitz <lungwitz@gmail.com>
Date:   Mon, 3 Jun 2019 00:04:20 +0200

    Add generic PathGenerator, InvCumulativeMersenneTwisterPathGenerator.

 SWIG/montecarlo.i    | 61 ++++++++++++++++++++++------------------------------
 SWIG/randomnumbers.i |  5 +++++
 2 files changed, 31 insertions(+), 35 deletions(-)

commit 2da9916a68524d38a1819b3352d5f3efc55e5a45
Author: Matthias Lungwitz <lungwitz@gmail.com>
Date:   Sun, 2 Jun 2019 21:18:13 +0200

    Fix Swaption Inheritance.

 SWIG/swaption.i | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

commit 24c2c708d0d5bd74667969013f08a59bce5f5fd8
Author: Matthias Lungwitz <lungwitz@gmail.com>
Date:   Fri, 31 May 2019 23:26:12 +0200

    Add SwaptionVolatilityDiscrete.

 SWIG/volatilities.i | 25 ++++++++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

commit d774d11755b117fc200424e57c08b04a787134a2

quantlib_swig_bindings/R/demo/bonds.R
## bonds.R -- following what bonds.py does for the QuantLib bindings for Python

suppressMessages(library(QuantLib))

## global data
calendar <- TARGET()

settlementDate <- Date(18, "September", 2008)
settlementDate <- Calendar_adjust(calendar, settlementDate)

fixingDays <- 3
settlementDays <- 3
todaysDate <- Calendar_advance(calendar, settlementDate, -fixingDays, "Days")
invisible(Settings_instance()$setEvaluationDate(d=todaysDate))

cat('Today          : ', todaysDate$`__str__`(), "\n")
cat('Settlement Date: ', settlementDate$`__str__`(), "\n")

## market quotes
## constructing bond yield curve
zcQuotes <- list(rates=c(0.0096, 0.0145, 0.0194),
                 tenor=c(Period(3, "Months"),
                 Period(6, "Months"),
                 Period(1, "Years")))

zcBondsDayCounter <- Actual365Fixed()

bondInstruments <- RateHelperVector()

for (i in 1:3) {
    r <- zcQuotes[["rates"]][i]
    tenor <- zcQuotes[["tenor"]][[i]]
    drh <- DepositRateHelper(QuoteHandle(SimpleQuote(r)),
                             tenor,
                             fixingDays,
                             calendar,
                             "ModifiedFollowing",
                             TRUE,
                             zcBondsDayCounter)
    RateHelperVector_push_back(bondInstruments, drh)
}

## setup bonds
redemption <- 100.0
numberOfBonds <- 5

bondQuotes <- list(list(Date(15,"March",2005),
                        Date(31,"August",2010),
                        0.02375, 100.390625),
                   list(Date(15,"June",2005),
                        Date(31,"August",2011),
                        0.04625, 106.21875),
                   list(Date(30,"June",2006),
                        Date(31,"August",2013),
                        0.03125, 100.59375),
                   list(Date(15,"November",2002),
                        Date(15,"August",2018),
                        0.04000, 101.6875),
                   list(Date(15,"May",1987),
                        Date (15,"May",2038),
                        0.04500, 102.140625)
                   )

# Definition of the rate helpers
for (i in 1:5) {
    issueDate <- bondQuotes[[i]][[1]]
    maturity <- bondQuotes[[i]][[2]]
    couponRate <- bondQuotes[[i]][[3]]
    marketQuote <- bondQuotes[[i]][[4]]
    schedule <- Schedule(issueDate, maturity, Period("Semiannual"),
                         UnitedStates("GovernmentBond"),
                         "Unadjusted", "Unadjusted",
                         copyToR(DateGeneration(), "Backward"),
                         FALSE)

    bh <- FixedRateBondHelper(QuoteHandle(SimpleQuote(marketQuote)),
                              settlementDays,
                              100.0,
                              schedule,
                              couponRate,
                              ActualActual("Bond"),
                              "Unadjusted",
                              redemption,
                              issueDate)
    RateHelperVector_push_back(bondInstruments, bh)
}
termStructureDayCounter <-  ActualActual("ISDA")

# not needed as defined in the interface file:  tolerance = 1.0e-15

bondDiscountingTermStructure <- PiecewiseFlatForward(settlementDate,
                                                     bondInstruments,
                                                     termStructureDayCounter)


# Building of the Libor forecasting curve
# deposits
dQuotes <- list(list(0.043375,  Period(1,"Weeks")),
                list(0.031875,  Period(1,"Months")),
                list(0.0320375, Period(3,"Months")),
                list(0.03385,   Period(6,"Months")),
                list(0.0338125, Period(9,"Months")),
                list(0.0335125, Period(1,"Years")))
sQuotes <- list(list(0.0295, Period(2,"Years")),
                list(0.0323, Period(3,"Years")),
                list(0.0359, Period(5,"Years")),
                list(0.0412, Period(10,"Years")),
                list(0.0433, Period(15,"Years")))

## deposits
depositDayCounter <- Actual360()
depoSwapInstruments <- RateHelperVector()

for (i in 1:length(dQuotes)) {
    rate  <- dQuotes[[i]][[1]]
    tenor <- dQuotes[[i]][[2]]
    drh <- DepositRateHelper(QuoteHandle(SimpleQuote(rate)),
                             tenor, fixingDays,
                             calendar, "ModifiedFollowing",
                             TRUE, depositDayCounter)
    RateHelperVector_push_back(depoSwapInstruments, drh)
}


## swaps
swFixedLegFrequency <- "Annual"
swFixedLegConvention <- "Unadjusted"
swFixedLegDayCounter <- Thirty360("European")
swFloatingLegIndex <- Euribor6M()
forwardStart <- Period(1,"Days")
for (i in 1:length(sQuotes)) {
    rate  <- sQuotes[[i]][[1]]
    tenor <- sQuotes[[i]][[2]]
    srh <- SwapRateHelper(QuoteHandle(SimpleQuote(rate)), tenor,
                          calendar, swFixedLegFrequency,
                          swFixedLegConvention, swFixedLegDayCounter,
                          swFloatingLegIndex, QuoteHandle(),forwardStart)
    RateHelperVector_push_back(depoSwapInstruments, srh)
}

depoSwapTermStructure <- PiecewiseFlatForward(settlementDate, depoSwapInstruments,
                                              termStructureDayCounter)

## Term structures that will be used for pricing:
## the one used for discounting cash flows

discountingTermStructure <- RelinkableYieldTermStructureHandle()

## the one used for forward rate forecasting
forecastingTermStructure <- RelinkableYieldTermStructureHandle()


########################################
##        BONDS TO BE PRICED           #
########################################

## common data

faceAmount <- 100

## pricing engine
bondEngine <- DiscountingBondEngine(discountingTermStructure)

## zero coupon bond
zeroCouponBond <- ZeroCouponBond(settlementDays,
                                 UnitedStates("GovernmentBond"),
                                 faceAmount,
                                 Date(15,"August",2013),
                                 "Following",
                                 116.92,
                                 Date(15,"August",2003))

invisible(Instrument_setPricingEngine(zeroCouponBond, bondEngine))

## fixed 4.5% US Treasury note

fixedBondSchedule <- Schedule(Date(15, "May", 2007),
                              Date(15, "May",2017), Period("Semiannual"),
                              UnitedStates("GovernmentBond"),
                              "Unadjusted", "Unadjusted",
                              copyToR(DateGeneration(), "Backward"), FALSE)

fixedRateBond <- FixedRateBond(settlementDays,
                               faceAmount,
                               fixedBondSchedule,
                               0.045,
                               ActualActual("Bond"),
                               "ModifiedFollowing",
                               100.0, Date(15, "May", 2007))
invisible(Instrument_setPricingEngine(fixedRateBond, bondEngine))

## Floating rate bond (3M USD Libor + 0.1%)
## Should and will be priced on another curve later...

liborTermStructure <- RelinkableYieldTermStructureHandle()

libor3m <- USDLibor(Period(3,"Months"),liborTermStructure)
invisible(Index_addFixing(libor3m, Date(17, "July", 2008), 0.0278625))

floatingBondSchedule <- Schedule(Date(21, "October", 2005),
                                 Date(21, "October", 2010), Period("Quarterly"),
                                 UnitedStates("NYSE"),
                                 "Unadjusted", "Unadjusted",
                                 copyToR(DateGeneration(), "Backward"), TRUE)

floatingRateBond <- FloatingRateBond(settlementDays,
                                     faceAmount,
                                     floatingBondSchedule,
                                     libor3m,
                                     Actual360(),
                                     "ModifiedFollowing",
                                     2,
                                     1.0,         # Gearings
                                     0.001,       # Spreads
                                     numeric(0),  #[],      # Caps
                                     numeric(0),  #[],      # Floors
                                     TRUE,    # Fixing in arrears
                                     100.0,
                                     Date(21, "October", 2005))

invisible(Instrument_setPricingEngine(floatingRateBond, bondEngine))

## coupon pricers

pricer <- BlackIborCouponPricer()

## optionlet volatilities
volatility <- 0.0
vol <- ConstantOptionletVolatility(settlementDays,
                                   calendar,
                                   "ModifiedFollowing",
                                   volatility,
                                   Actual365Fixed())

invisible(IborCouponPricer_setCapletVolatility(pricer, OptionletVolatilityStructureHandle(vol)))
invisible(setCouponPricer(Bond_cashflows(floatingRateBond), pricer))


## Yield curve bootstrapping
invisible(RelinkableQuoteHandle_linkTo(forecastingTermStructure, depoSwapTermStructure))
invisible(RelinkableQuoteHandle_linkTo(discountingTermStructure, bondDiscountingTermStructure))

## We are using the depo & swap curve to estimate the future Libor rates
invisible(RelinkableQuoteHandle_linkTo(liborTermStructure, depoSwapTermStructure))

##
df <- data.frame(zeroCoupon=c(Instrument_NPV(zeroCouponBond),
                   Bond_cleanPrice(zeroCouponBond),
                   Bond_dirtyPrice(zeroCouponBond),
                   Bond_accruedAmount(zeroCouponBond),
                   NA,
                   NA,
                   100*Bond_yield(zeroCouponBond, Actual360(), "Compounded", "Annual")),
                 fixedRate=c(Instrument_NPV(fixedRateBond),
                   Bond_cleanPrice(fixedRateBond),
                   Bond_dirtyPrice(fixedRateBond),
                   Bond_accruedAmount(fixedRateBond),
                   100*Bond_previousCouponRate(fixedRateBond),
                   100*Bond_nextCouponRate(fixedRateBond),
                   100*Bond_yield(fixedRateBond, Actual360(), "Compounded", "Annual")),
                 floatingRate=c(Instrument_NPV(floatingRateBond),
                   Bond_cleanPrice(floatingRateBond),
                   Bond_dirtyPrice(floatingRateBond),
                   Bond_accruedAmount(floatingRateBond),
                   100*Bond_previousCouponRate(floatingRateBond),
                   100*Bond_nextCouponRate(floatingRateBond),
                   100*Bond_yield(floatingRateBond, Actual360(), "Compounded", "Annual")),
                 row.names=c("NPV", "Clean Price", "Dirty Price",
                 "Accrued Amount", "Previous Coupon", "Next Coupon", "Yield"))
cat("\nResults:\n")
print(df, digits=5)

# Other computations

cat("\nSample indirect computations (for the floating rate bond):\n")
yld <- Bond_yield(floatingRateBond, Actual360(), "Compounded", "Annual")
clnPrc <- Bond_cleanPrice(floatingRateBond, yld, Actual360(), "Compounded", "Annual", settlementDate)
cat("Yield to Clean Price: ", clnPrc, "\n")
yld <- Bond_yield(floatingRateBond, clnPrc, Actual360(), "Compounded", "Annual",settlementDate)
cat("Clean Price to Yield: ", yld, "\n")
